OC4IDS_Database_Data_Import.ipynb

```diff
@@ -286,6 +300,46 @@
         "  with open('project_package.json', 'wb') as f:\n",
         "    f.write(response.content)\n",
         "\n",
+        "elif source_id == 'uganda_gpp':\n",
+        "    packages = []\n",
+        "    start_year = 2022\n",
+        "    current_year = datetime.now().year\n",
+        "    years = list(range(start_year, current_year + 1))\n",
+        "\n",
+        "    for year in years:\n",
+        "      url = f\"{sources[source_id]}{year}\" # Construct the url\n",
+        "      response = requests.get(url, verify=False, stream=True) # stream=True for large files\n",
+        "      zip_data = io.BytesIO(response.content)\n",
+        "\n",
+        "      try:\n",
+        "          with zipfile.ZipFile(zip_data, 'r') as zip_ref:\n",
+        "              for filename in zip_ref.namelist(): #namelist returns all the filepaths in the zip file\n",
+        "                  if filename.endswith('.json'):\n",
+        "                      with zip_ref.open(filename) as f:\n",
+        "                          package = json.load(f)\n",
+        "                          packages.append(package)\n",
+        "      except zipfile.BadZipFile:\n",
+        "          print(url, \"Error: The downloaded file is not a valid zip archive.\")\n",
+        "\n",
+        "\n",
+        "    # Combine the packages\n",
+        "    combined_package = combine_project_packages(\n",
+        "      packages,\n",
+        "      uri=\"https://gpp.ppda.go.ug/public/open-data/oc4ids/project-datasets\",\n",
+        "      publisher={\n",
+        "        \"name\": \"The Public Procurement and Disposal of Public Assets Authority\",\n",
+        "        \"scheme\": \"PPDA\",\n",
+        "        \"uid\": \"475\",\n",
+        "        \"uri\": \"https://www.ppda.go.ug/\"\n",
+        "      },\n",
+        "      published_date = datetime.now().strftime('%Y-%m-%dT%H:%M:%S+00:00'),#Extract date from files?\n",
+        "      version=\"0.9\"  #Extract version from the latest file?\n",
```

Comment:
Yep, I think we should extract this from the data, rather than hardcode it as it affects how the data is validated.

Before combining the packages, I suggest checking that they all set the same version, e.g.

```python
versions = set([package["version"] for package in packages])
if len(versions) > 1:
   raise ValueError(f"Packages declare more than one version")
```

Then set `version` to `versions[0]`.

The publisher data is less important because it doesn't affect validation, but probably still good to set it from the final package, rather than hardcode it.

Reply:
Thanks Duncan - I had forgotten all about sets! It's ready for review now